reading-notes-intro-computation-programming-python/chapter-two_programming-basics/chapter-2.md

```diff
@@ -222,36 +391,90 @@ def birthday():
     print(f'You were born in the year {birthday[6:10]}')
 ```
 
-#### operators, overloading
+### operate on text
 
-> The operator + is said to be overloaded because it has different meanings depending upon the types of objects to which it is applied.
+##### length
 
-operator +:
-- type number: addition
-- type string: concatenation
+The `len` function returns the length (character count) of a string. 
 
-operator * :
-- type int: multiplication / product
-- type string: repetition 
+```python
+len('abc') == 3 # True
+```
+
+##### extracting
+
+As the string is [non-scalar](#type) type, you can access its individual parts.
+
+It is a `zero-based` array :
+- text start at `index` 0;
+- text end at index (length - 1).
+
+To access a character, use the bracket operator `[index]`:
+- if an index is out of range, you get an error;
+- if the index is negative `[-n]`, you get the nth last character, which is (length - n) index.
 
 ```python
-3 * "a" == 'aaa' # True
+'abc'[0] == 'a'   # True
+'abc'[3]          # IndexError: string index out of range
+'abcd'[-2] == 'c' # True
+'abc'[-3] == 'a'  # True
+'abc'[-4]         # IndexError: string index out of range
 ```
 
-#### operator, functions
+To access a sequence of character, aka a substring, use the slice operator.
+The slice operator is the bracket operator with two parameters : `[start:end]`.
 
-##### length
+The selection :
+- begins at `start` index (lower bound), including; 
+- stop at `end` index (upper bound), excluding - which may be surprising.
 
-The `len` function returns the length (character count) of a string. 
+The matching mathematics interval is `[start:end[`.
 
 ```python
-len('abc') == 3 # True
+'abc'[1:2] == 'c'  # True
+'abc'[1:3] == 'bc' # True
+'abc'[0:len('abc')] == 'abc' # True
 ```
 
-####
-indexing : 'abc'[0] -> 'a'/ Error if out of range like 'abc'[10] / 'abc'[-1] -> 'c'
-slicing : abc'[0:len('abc')] -> 'abc' / 'abc'[:] -> 'abc' (if values are omited around colon, defaults values are 0 and string length)
- 
+If values are omitted around colon, defaults values are : 
+- start : 0;
+- end : string length.
+
+```python
+'abc'[:] == 'abc' # True
+``` 
+
+// Todo: check what's happening, as it seems no copy is performed
+
+This is a quick way to take a copy of the object
+[Source](https://towardsdatascience.com/mastering-indexing-and-slicing-in-python-443e23457125/)
+
+```python
+id('abc') == id('abc'[:]) # False
+```
+
+#### concatenating, combining
+
+Use the `+` operator.
+```python
+'hello' + ' ' + 'world' == 'hello world' # True
+```
+
+The concatenation returns a new object. 
+[Source](https://docs.python.org/3/library/stdtypes.html#common-sequence-operations)
+> Concatenating immutable sequences always results in a new object. This means that building up a sequence by repeated concatenation will have a quadratic runtime cost in the total sequence length. To get a linear runtime cost, you must switch to one of the alternatives below
```

Comment:
Thanks for the enlightenment about mutability/immutability and the costs on runtime it generates